lib/src/table/table.service.ts

```diff
@@ -0,0 +1,386 @@
+
+        return this._sortInfo;
+    }
+    private set sortInfo(value: SortInfo) {
+        this._sortInfo = { ...value };
+    }
+
+    @Output() currentSortInfo: BehaviorSubject<SortInfo<keyof T>> = new BehaviorSubject(null);
+    @Output() isAllSelected: BehaviorSubject<boolean> = new BehaviorSubject(false);
+
+    private columnsList: string[];
+
+    /** The Master Table, Sorted */
+    private _sortedTable: T[] = [];
+    @Output() sortedTable: BehaviorSubject<T[]> = new BehaviorSubject([]);
+    /** The Master Table, Paginated */
+    private _paginatedTable: T[][] = [[]];
+    @Output() paginatedTable: BehaviorSubject<T[][]> = new BehaviorSubject([[]]);
+    /** The indexes of the selected rows, paginated */
+    private _selectedRows: number[][] = [[]];
+    @Output() selectedRows: BehaviorSubject<number[][]> = new BehaviorSubject([[]]);
+    /** The Currently Visible Page of the Sorted and Paginated Master Table */
+    private _currentTable: T[] = [];
+    @Output() currentTable: BehaviorSubject<T[]> = new BehaviorSubject([]);
+    /** The Index of the currently selected page */
+    private _currentPageIndex: number = 0;
+    @Output() currentPageIndex: BehaviorSubject<number> = new BehaviorSubject(0);
+    /** The Header List of The Master Table with all the Column Info */
+    private _tableHeaderList: Array<TableHeaderListItem<T>> = [];
+    @Output() tableHeaderList: BehaviorSubject<TableHeaderListItem<T>[]> = new BehaviorSubject([]);
+
+    // ------------- CONSTRUCTOR ----------------
+    constructor() {}
+
+    public registerDatasource(table: T[], config: TableConfig<T> = {}) {
+        this.table = table;
+        this.tableConfig = config;
+        this.reloadTable();
+    }
+
+    // ------------- HELPERS --------------------
+    /**
+     * INIT CONFIG TYPES
+     * @returns the initial configuration for config types
+     */
+    private initConfigTypes(): TableConfig<T>["types"] {
+        if (this.table && this.table.length) {
+            const types: TableConfig<T>["types"] = this.table.reduce((result, item) => {
+                for (const key in item) {
+                    if (item[key]) {
+                        const type = (typeof item[key]).toLowerCase();
+                        if (type === "string" || type === "date" || type === "number") {
+                            result[key.toString()] = type;
+                        } else {
+                            result[key.toString()] = "string";
+                        }
+                    }
+                }
+                return result;
+            }, {});
+            return types;
+        }
+
+        return {};
+    }
+
+    /**
+     * RELOAD TABLE:
+     * Responsible for recalculation (or initialisation) of everything needed for the table including the sort and pagination and rebuilds all the table components
+     */
+    private reloadTable(): void {
+        const table: T[] = this.table && this.table.length ? [...this.table] : [];
```

Comment:
will there be multiple level for the data? if there is, we might need to deep clone the array

Reply:
> can we also add example for custom header?

@kherP like a header which has custom html? I don't have that feature yet but I can add it if you like.

Reply:
for now no multiple levels are supported yet so this should be fine for now